### PR TITLE
永続化 migration / import-export の property-based test を追加 (#718)

### LINKED ISSUE #718: storage migration / import-export に property-based test と malformed-data fuzz test を追加する
Parent: #724

Cross-cutting validation for: #712, #713, #728, #730, #732

## 優先度

P1

## 背景

TABBINのstorage / import-exportはURL参照、旧形式migration、custom project order、category relationなど組み合わせが多く、example-based testだけではedge caseを網羅しづらい。

Persistence Model v2 / IndexedDB移行では `Url` / `Collection` / `CollectionMembership` / `CollectionCategory` / `CollectionGroup` のrelation組み合わせ、legacy storage mapping、Backup V2 migration、timestamp fallback等により状態空間がさらに増える。

このIssueは#724配下のcross-cutting test taskとして、migration / import-export / integrity checkerのinvariantをpropertyとして検証する。

## 対象候補

- URL normalization
- Persistence Model v2 invariant
- #712 integrity checker
- #728 legacy chrome.storage → v2 migration
- #713 backup schema migration
- #730 import / export
- #732 Analytics timestamp semanticsに関係するmigration output
- category / collection order normalization

### normalization idempotence

```text
normalize(normalize(x)) === normalize(x)
```

### current migration idempotence

```text
migrateToCurrent(current) preserves current semantics
```

### legacy storage migration determinism

```text
migrateLegacy(x, fixedClock, fixedIdGenerator)
===
migrateLegacy(x, fixedClock, fixedIdGenerator)
```

### integrity checker determinism

```text
check(snapshot) === check(snapshot)
```

### safe repair idempotence

```text
repair(repair(x)) === repair(x)
```

repairable operationに限定する。

### backup round trip

```text
import(export(v2)) preserves required invariants
```

完全なobject equalityではなく、仕様上preserveすべきlogical data / relation / timestampを定義して比較する。

### migration preserves relation

```text
legacy URL membership
→ v2 Membership exists

legacy notes
→ Membership.notes preserved

legacy category assignment
→ Membership.categoryId points to matching CollectionCategory
```

### timestamp relation

```text
Url.firstSavedAt <= Url.lastSavedAt
```

migration fallback後も成立する。

## generator / arbitrary

`fast-check`等のproperty-based testing library採用可否を確認する。

採用する場合、generatorをtest fileごとに重複定義しない。

候補:

```text
src/test/arbitraries/persistence/
  urlArbitrary.ts
  collectionArbitrary.ts
  membershipArbitrary.ts
  persistenceSnapshotArbitrary.ts
  legacyStorageSnapshotArbitrary.ts
  backupArbitrary.ts
```

DDD / architecture ruleに合うactual pathは実装時に調整する。

## v2 model generator

生成対象:

- valid Url
- valid Collection domain/custom
- valid Membership
- valid Category
- valid Group
- valid PersistenceV2Snapshot

valid generatorはinvariantを満たすdataを生成する。

invalid generatorは意図的に1つまたは複数invariantを破る。

## malformed / corrupted data generator

対象例:

- duplicate URL ID
- duplicate normalized URL
- dangling membership URL
- dangling membership Collection
- cross-collection category reference
- missing group target
- duplicate membership
- duplicate canonical domain collection
- invalid sortOrder
- `firstSavedAt > lastSavedAt`
- empty / whitespace / unusual Unicode
- extreme timestamp
- mixed legacy/current backup fields
- nested `urls` + `urlIds` mismatch
- duplicate parent relation sources
- unknown additional fields
- malformed schemaVersion

## legacy storage generator

#728のactual legacy schemaを利用する。

以下の歴史的shapeを生成できるようにする。

- `TabGroup.urls`中心
- `TabGroup.urlIds`中心
- `urls` + nested URLs混在
- `CustomProject.urls`
- `CustomProject.urlIds`
- `urlMetadata`
- `urlSubCategories`
- `ParentCategory.domains`
- `domainNames`
- `parentCategoryId`
- `DomainParentCategoryMapping`

current production typeをlegacy arbitraryのtypeとして再利用しない。

## seed reproduction

failure seedを必ず出力し、再現可能にする。

例:

```text
Property failed after 134 tests
seed: 18927364
path: "12:3:1"
```

CI failureからlocalで同じcaseを再実行できるcommand / docsを用意する。

## regression fixture昇格

property testで見つかったbugをrandom generatorだけに任せ続けない。

以下の場合はexplicit regression fixture / example testへ昇格する。

- production data loss risk
- migration conflict rule bug
- timestamp corruption
- note / category relation loss
- source-of-truth cutover bug
- legacy backup compatibility bug

## execution strategy

通常quality gateを過度に遅くしない。

候補:

```text
PR quality
→ deterministic property tests with moderate runs

nightly / manual
→ higher numRuns / broader malformed cases
```

critical migration propertyはPR gateに含める。

「重いからmigration test全体をnightlyだけ」にしない。

## #712 integrity checkerとの関係

invalid snapshot generatorでissue typeを検証する。

例:

```text
generate valid snapshot
  ↓
remove referenced Url
  ↓
check
  ↓
URL_MISSING exactly detected
```

複数corruptionを生成する場合、expected issue集合を検証する。

## #728 live migrationとの関係

legacy snapshotを生成し、migration resultに対して:

- #712 healthy
- semantic preservation
- deterministic mapping
- timestamp invariant

を検証する。

ID generationはfixed / deterministic adapterを利用する。

## #730 backupとの関係

Backup V2 round trip propertyを追加する。

pre-IDB legacy backup property testは2026年8月31日までのtemporary compatibility scope。

#734 cleanup後はlegacy importer property testを削除し、archived historical fixtureとcurrent Backup V2 testsを分離する。

## #732 Analyticsとの関係

migration outputで:

```text
firstSavedAt <= lastSavedAt
```

を保証する。

same URLがmultiple Collectionへ所属するgeneratorを作り、Membership.addedAtがUrl.lastSavedAtへ不正同期されないことを検証する。

## 受け入れ条件

- [ ] property-based testingの対象invariantが明文化されている
- [ ] Persistence Model v2 valid / invalid generatorがある
- [ ] #712 integrity checkerにproperty testがある
- [ ] #728 legacy migrationにdeterminism / semantic preservation propertyがある
- [ ] migration timestamp invariantを検証している
- [ ] malformed / partially corrupted data generatorがある
- [ ] failure seedを再現できる
- [ ] critical generated failureをregression fixtureへ昇格する基準がある
- [ ] Backup V2 round trip propertyを検証している
- [ ] pre-IDB legacy backup property testが#734 cleanup対象として識別できる
- [ ] test runtimeが通常quality gateを過度に悪化させない

## 注意点

- random testを非決定的に実行するだけにしない。seed再現を必須にする
- property数を増やす前に#725 / #712でdomain invariantを定義する
- coverage数値を上げるためではなく、状態空間の組み合わせbug検出を目的にする
- current production typeへlegacy optional fieldを戻すためのtest helperを作らない

### PR DESCRIPTION
**Pull request by haruto-yamada1**
## 原因 / 背景

Issue #718。storage / import-export は URL 参照・旧形式 migration・custom project order・
category relation の組み合わせが多く、example-based test だけでは edge case を網羅しづらい。
#724 配下の cross-cutting task として、migration / import-export / integrity checker の
invariant を property として検証する基盤を追加する。

## 採用した解決

fast-check を devDependency に追加し、共有 arbitrary を
`src/test/arbitraries/persistence/` に集約（test file ごとの重複定義を防止）。
対象はすべて純粋関数（`mapLegacyStorageToPersistenceV2`、`checkPersistenceIntegrity`、
`createStorageRepairPlan`、`BackupMapper`、`convertLegacyBackup`、URL/domain
normalization）なので、clock / id generator の stub なしに決定性を直接検証できる。

## 主要変更

- `src/test/arbitraries/persistence/`
  - `primitives.ts`: timestamp / sortOrder / domain / URL / display text
  - `persistenceSnapshotArbitrary.ts`: v2 valid snapshot generator + 10 種の
    corruption generator（誘発 issue code 付き）
  - `legacyStorageSnapshotArbitrary.ts`: 歴史的 legacy shape（urlIds / nested /
    parallel mixed、urlSubCategories、urlMetadata、ParentCategory、mappings）の
    well-formed generator + malformed raw generator
  - `backupArbitrary.ts`: canonical user settings + pre-IDB backup envelope
  - `fastCheckParameters.ts`: `FAST_CHECK_SEED` / `FAST_CHECK_RUNS` による
    seed 再現・run 数制御
- Property tests（6 file / 19 tests）
  - normalization idempotence
  - #712 integrity checker の決定性 + corruption 検出
  - safe repair の fixpoint（repair(repair(x)) === repair(x)）
  - #728 legacy migration の決定性・semantic preservation・timestamp invariant・
    #732 shared-URL membership semantics
  - #713/#730 Backup V2 round trip
  - pre-IDB legacy backup 変換（#734 cleanup 対象として明示）
- `docs/testing/property-based-tests.md`: 対象 invariant の明文化、seed 再現手順、
  regression fixture 昇格基準、execution strategy、#734 識別
- `.oxlintrc.json`: `vitest/expect-expect` の `assertFunctionNames` に
  `expectTypeOf` / `fc.assert` を追加（property test を assertion として認識）

実装中に generator 側の学習として、legacy mixed shape は `urlIds` と nested `urls` の
並列表現（id・url・title・savedAt 完全一致）のみ conflict-free であり、CustomProject は
`urlIds` 存在時に同長の `urls` を必須とする mapper 仕様を確認。production code の変更は
ない（test-only）。

## 受け入れ条件への対応

- [x] 対象 invariant が明文化（`docs/testing/property-based-tests.md` の対応表）
- [x] v2 valid / invalid generator（`persistenceSnapshotArbitrary.ts`）
- [x] #712 integrity checker に property test
- [x] #728 legacy migration に determinism / semantic preservation property
- [x] migration timestamp invariant（`firstSavedAt <= lastSavedAt` 等、malformed 入力でも検証）
- [x] malformed / partially corrupted data generator
- [x] failure seed 再現（`FAST_CHECK_SEED`、docs に手順）
- [x] regression fixture 昇格基準（docs に明記）
- [x] Backup V2 round trip property
- [x] pre-IDB legacy backup property test を #734 cleanup 対象として識別
  （file 先頭コメント + docs）
- [x] runtime: 19 tests / 約 0.5 秒（`FAST_CHECK_RUNS=200` でも）で quality gate を悪化させない

## 検証結果

- 新規 property tests: 19 tests pass（既定 50 runs、確認用に 200 runs でも pass）
- `FAST_CHECK_SEED=18927364` での seed replay 動作確認
- `bun run quality:check`: pass（format / lint / compile / 409 test files /
  knip / duplication / secretlint / arch:check）
- `bun run test:coverage`: pass（global 96.14% stmts / 91% branches /
  96.96% funcs / 96.12% lines、全 threshold 充足）

## Regression risk

低。test-only の追加で production code 変更なし。`.oxlintrc.json` の
`assertFunctionNames` 拡張は既存テスト（expect / expectTypeOf）の判定を維持したまま
`fc.assert` を追加するのみ。coverage は import-export critical glob を含め全 threshold を
満たす。

Closes #718

## Summary by CodeRabbit

* **テスト**
  * 保存データの変換、整合性検証、修復計画、URL正規化、バックアップ入出力を対象としたプロパティベーステストを追加しました。
  * 正常・異常データを用いて、決定性、データ保持、不整合の検出・修復を幅広く検証します。
  * CIで共通テストも実行されるようになりました。
* **ドキュメント**
  * テスト方針、再現手順、実行戦略、回帰テストへの昇格基準を追加しました。
* **開発者向け改善**
  * テスト実行回数やシードを設定でき、失敗ケースを再現しやすくなりました。

### CODE REVIEW

File: src/test/arbitraries/persistence/fastCheckParameters.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+): number | undefined => {
+  if (value === undefined) {
+    return undefined
+  }
+  const parsed = Number(value)
+  return Number.isSafeInteger(parsed) && parsed > 0 ? parsed : undefined
+}
+
+const numRuns = parsePositiveInteger(process.env.FAST_CHECK_RUNS)
+const seed = parsePositiveInteger(process.env.FAST_CHECK_SEED)
```
[the remark this answers is not in the document]
Comment (haruto-yamada1) [reply src/test/arbitraries/persistence/fastCheckParameters.ts]: 対応しました (commit 7ac23711)。

**修正内容**
- `FAST_CHECK_SEED` のパースを `parseSafeInteger` に分離し、負の整数と `0` を受理するように変更（`numRuns` は従来どおり `parsePositiveInteger` で正の整数のみ）。
- 空文字 / 空白のみ / 非整数 / 非 safe integer は従来どおり `undefined`（seed 省略）扱い。
- ご指摘のとおり fast-check は失敗時に負の seed を報告するため、replay 手順が壊れていました。`85c1ed9` 時点のコードで RED（negative / zero seed が drop される）を確認後に修正しています。

**検証**
- 新規 regression test: `src/test/arbitraries/persistence/fastCheckParameters.test.ts`（negative / zero seed の保持、無効値の drop、numRuns の正整数制約を網羅）
- `bunx vitest run --config vitest.ci.config.ts --project node src/test/arbitraries/persistence/...` → 26 tests passed
- `bun run quality:check` → 全ゲート通過
- `bun run test:coverage` → 全 threshold 通過（global 96.15% stmts ほか）

なお `src/test/**` が Vitest project の include 対象外で新規テストが CI で実行されなかったため、`vitest.ci.config.ts` の node project に `src/test/**/*.test.ts` を追加しています。

File: src/test/arbitraries/persistence/fastCheckParameters.test.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  vi.unstubAllEnvs()
+  if (env.runs !== undefined) {
+    vi.stubEnv('FAST_CHECK_RUNS', env.runs)
+  }
+  if (env.seed !== undefined) {
+    vi.stubEnv('FAST_CHECK_SEED', env.seed)
+  }
+  const { fastCheckParameters } = await import('./fastCheckParameters')
+  return fastCheckParameters
+}
```
[the remark this answers is not in the document]
Comment (haruto-yamada1) [reply src/test/arbitraries/persistence/fastCheckParameters.test.ts:17]: 対応しました (commit 2fc5f40d)。

**修正内容**
- `importParameters` で `vi.unstubAllEnvs()` 直後に `FAST_CHECK_RUNS` / `FAST_CHECK_SEED` を `undefined` へ明示的に stub し、その後で引数の値だけを上書き。外部環境の値が各 import に継承されないようにしました。

**検証**
- 修正前: `FAST_CHECK_SEED=123 FAST_CHECK_RUNS=200 bunx vitest run ...` で RED（外部 env を継承して失敗）を再現
- 修正後: 同コマンドで 11 tests passed、外部 env なしでも 11 tests passed
- `bun run quality:check` → 全ゲート通過
